platforms/documentation/docs/src/docs/userguide/dep-man/dependency-management/resolution_rules.adoc

```diff
@@ -0,0 +1,474 @@
+==== Substituting a dependency with a dependency with capabilities
+
+you can substitute a dependency with a different variant that includes specific **capabilities**.
+Capabilities allow you to specify that a particular variant of a dependency offers a set of related features or functionality, such as <<java_testing.adoc#sec:java_test_fixtures,test fixtures>>.
+
+This example substitutes a regular dependency with its <<java_testing.adoc#sec:java_test_fixtures,test fixtures>> using a capability:
+
+====
+include::sample[dir="snippets/dependencyManagement/customizingResolution-capabilitySubstitutionRule/kotlin",files="build.gradle.kts[tags=substitution_rule]"]
+include::sample[dir="snippets/dependencyManagement/customizingResolution-capabilitySubstitutionRule/groovy",files="build.gradle[tags=substitution_rule]"]
+====
+
+Here, we substitute the regular `com.acme:lib:1.0` dependency with its `lib-test-fixtures` variant.
+The `requireCapability` function specifies that the new variant must have the `com.acme:lib-test-fixtures` capability, ensuring the right version of the dependency is selected for testing purposes.
+
+Capabilities within the substitution rule are used to precisely match dependencies, and Gradle only substitutes dependencies that match the required capabilities.
+
+Refer to the link:{javadocPath}/org/gradle/api/artifacts/DependencySubstitutions.html#variant-org.gradle.api.artifacts.component.ComponentSelector-org.gradle.api.Action-[DependencySubstitutions API] for a complete reference of the variant substitution API.
+
+[[sec:substitution_with_classifier]]
+=== Substituting a dependency with a classifier or artifact
+
+You can substitute dependencies that have a classifier with ones that don't or vice versa. Classifiers are often used to represent different versions of the same artifact, such as platform-specific builds or dependencies with different APIs.
+Although Gradle discourages the use of classifiers, it provides a way to handle substitutions for cases where classifiers are still in use.
+
+Consider the following setup:
+
+====
+include::sample[dir="snippets/dependencyManagement/customizingResolution-classifierSubstitutionRule/kotlin",files="consumer/build.gradle.kts[tags=dependencies]"]
+include::sample[dir="snippets/dependencyManagement/customizingResolution-classifierSubstitutionRule/groovy",files="consumer/build.gradle[tags=dependencies]"]
+====
+
+In the example above, the first level dependency on `quasar` makes us think that Gradle would resolve `quasar-core-0.8.0.jar` but it's not the case.
+
+The build fails with this message:
+
+[source,text]
+----
+Execution failed for task ':resolve'.
+> Could not resolve all files for configuration ':runtimeClasspath'.
+   > Could not find quasar-core-0.8.0-jdk8.jar (co.paralleluniverse:quasar-core:0.8.0).
+     Searched in the following locations:
+         https://repo1.maven.org/maven2/co/paralleluniverse/quasar-core/0.8.0/quasar-core-0.8.0-jdk8.jar
+----
```

Comment:
🤔 Is this really the error message from this scenario?  This is awful, it doesn't mention capabilities being the reason at all!  If this is, please let me know and I'll add it to my list of things to look at, since it already has a reproducer scenario here.

Reply:
I have some bad news, it gave the error:

> Execution failed for task ':consumer:resolve'.
> > Could not resolve all files for configuration ':consumer:runtimeClasspath'.
>    > Could not find quasar-core-0.8.0-jdk8.jar (co.paralleluniverse:quasar-core:0.8.0).
>      Searched in the following locations:
>          https://repo.maven.apache.org/maven2/co/paralleluniverse/quasar-core/0.8.0/quasar-core-0.8.0-jdk8.jar
> 
> Possible solution:
>  - Declare repository providing the artifact, see the documentation at https://docs.gradle.org/current/userguide/declaring_repositories.html